File: src/main/java/com/entrecopas/financiero/service/FinancieroService.java
package com.entrecopas.financiero.service;

import com.entrecopas.exception.ResourceNotFoundException;
import com.entrecopas.financiero.dto.DashboardFinancieroDTO;
import com.entrecopas.financiero.dto.DesgloseLoteFinancieroDTO;
import com.entrecopas.financiero.dto.DesgloseProductoFinancieroDTO;
import com.entrecopas.financiero.dto.FinancieroKPIsDTO;
import com.entrecopas.lote.model.Lote;
import com.entrecopas.lote.repository.LoteRepository;
import com.entrecopas.producto.model.Producto;
import com.entrecopas.producto.repository.ProductoRepository;
import com.entrecopas.productor.model.Productor;
import com.entrecopas.productor.repository.ProductorRepository;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Servicio de negocio para el cálculo de métricas financieras, costos y mermas del productor.
 * Aplica aislamiento multi-tenant estricto sobre los datos contables.
 */
@Service
public class FinancieroService {

    private final ProductorRepository productorRepository;
    private final ProductoRepository productoRepository;
    private final LoteRepository loteRepository;

    public FinancieroService(ProductorRepository productorRepository,
                             ProductoRepository productoRepository,
                             LoteRepository loteRepository) {
        this.productorRepository = productorRepository;
        this.productoRepository = productoRepository;
        this.loteRepository = loteRepository;
    }

    /**
     * Calcula y consolida el dashboard financiero del productor autenticado.
     */
    @Transactional(readOnly = true)
    public DashboardFinancieroDTO obtenerDashboard(Long usuarioId) {
        return obtenerDashboard(usuarioId, null, null, null);
    }

    /**
     * Calcula y consolida el dashboard financiero del productor con filtros opcionales.
     */
    @Transactional(readOnly = true)
    public DashboardFinancieroDTO obtenerDashboard(Long usuarioId, Long productoId,
                                                   java.time.LocalDate fechaDesde,
                                                   java.time.LocalDate fechaHasta) {
        Productor productor = productorRepository.findByUsuarioId(usuarioId)
                .orElseThrow(() -> new ResourceNotFoundException(
                        "No se encontró una bodega o taller artesanal asociado a este usuario."));

        List<Producto> productos = productoRepository.findByProductorId(productor.getId());
        List<Lote> todosLotes = loteRepository.findByProductoProductorIdOrderByFechaProduccionDesc(productor.getId());

        List<Lote> lotes = todosLotes.stream()
                .filter(l -> productoId == null || (l.getProducto() != null && l.getProducto().getId().equals(productoId)))
                .filter(l -> fechaDesde == null || (l.getFechaProduccion() != null && !l.getFechaProduccion().isBefore(fechaDesde)))
                .filter(l -> fechaHasta == null || (l.getFechaProduccion() != null && !l.getFechaProduccion().isAfter(fechaHasta)))
                .collect(Collectors.toList());

        if (lotes.isEmpty()) {
            return new DashboardFinancieroDTO();
        }

        BigDecimal costoTotalAcumulado = BigDecimal.ZERO;
        BigDecimal volumenTotalLitros = BigDecimal.ZERO;
        BigDecimal mermaTotalLitros = BigDecimal.ZERO;

        List<DesgloseLoteFinancieroDTO> desgloseLotes = new ArrayList<>();

        for (Lote lote : lotes) {
            Producto prod = lote.getProducto();
            BigDecimal vol = lote.getVolumenLitros() != null ? lote.getVolumenLitros() : BigDecimal.ZERO;
            BigDecimal merma = lote.getMermaLitros() != null ? lote.getMermaLitros() : BigDecimal.ZERO;
            BigDecimal costo = lote.getCostoTotal() != null ? lote.getCostoTotal() : BigDecimal.ZERO;
            BigDecimal volNeto = vol.subtract(merma).max(BigDecimal.ZERO);

            costoTotalAcumulado = costoTotalAcumulado.add(costo);
            volumenTotalLitros = volumenTotalLitros.add(vol);
            mermaTotalLitros = mermaTotalLitros.add(merma);

            BigDecimal mermaPct = vol.compareTo(BigDecimal.ZERO) > 0
                    ? merma.multiply(BigDecimal.valueOf(100)).divide(vol, 2, RoundingMode.HALF_UP)
                    : BigDecimal.ZERO;

            BigDecimal costoPorLitro = volNeto.compareTo(BigDecimal.ZERO) > 0
                    ? costo.divide(volNeto, 2, RoundingMode.HALF_UP)
                    : BigDecimal.ZERO;

            double litrosPorUnidad = 0.75;
            if (prod != null && prod.getPresentacion() != null) {
                String pres = prod.getPresentacion().toLowerCase();
                if (pres.contains("330")) {
                    litrosPorUnidad = 0.33;
                } else if (pres.contains("500")) {
                    litrosPorUnidad = 0.50;
                }
            }

            int unidadesEstimadas = (int) (volNeto.doubleValue() / litrosPorUnidad);
            BigDecimal costoPorUnidad = unidadesEstimadas > 0
                    ? costo.divide(BigDecimal.valueOf(unidadesEstimadas), 2, RoundingMode.HALF_UP)
                    : BigDecimal.ZERO;

            BigDecimal precioVenta = prod != null && prod.getPrecio() != null ? prod.getPrecio() : BigDecimal.ZERO;
            BigDecimal ingresoBruto = precioVenta.multiply(BigDecimal.valueOf(unidadesEstimadas));
            BigDecimal gananciaBruta = ingresoBruto.subtract(costo);

            BigDecimal margenBrutoPct = ingresoBruto.compareTo(BigDecimal.ZERO) > 0
                    ? gananciaBruta.multiply(BigDecimal.valueOf(100)).divide(ingresoBruto, 1, RoundingMode.HALF_UP)
                    : BigDecimal.ZERO;

            desgloseLotes.add(new DesgloseLoteFinancieroDTO(
                    lote.getId(),
                    lote.getCodigoTrazabilidad(),
                    prod != null ? prod.getNombre() : "Bebida Artesanal",
                    prod != null ? prod.getPresentacion() : "750ml",
                    lote.getFechaProduccion(),
                    vol.setScale(2, RoundingMode.HALF_UP),
                    merma.setScale(2, RoundingMode.HALF_UP),
                    mermaPct,
                    volNeto.setScale(2, RoundingMode.HALF_UP),
                    costo.setScale(2, RoundingMode.HALF_UP),
                    costoPorLitro,
                    unidadesEstimadas,
                    costoPorUnidad,
                    precioVenta.setScale(2, RoundingMode.HALF_UP),
                    margenBrutoPct
            ));
        }

        BigDecimal volumenNetoTotal = volumenTotalLitros.subtract(mermaTotalLitros).max(BigDecimal.ZERO);
        BigDecimal porcentajeMermaGlobal = volumenTotalLitros.compareTo(BigDecimal.ZERO) > 0
                ? mermaTotalLitros.multiply(BigDecimal.valueOf(100)).divide(volumenTotalLitros, 2, RoundingMode.HALF_UP)
                : BigDecimal.ZERO;

        BigDecimal costoPromedioLitro = volumenNetoTotal.compareTo(BigDecimal.ZERO) > 0
                ? costoTotalAcumulado.divide(volumenNetoTotal, 2, RoundingMode.HALF_UP)
                : BigDecimal.ZERO;

        FinancieroKPIsDTO kpis = new FinancieroKPIsDTO(
                costoTotalAcumulado.setScale(2, RoundingMode.HALF_UP),
                volumenTotalLitros.setScale(2, RoundingMode.HALF_UP),
                mermaTotalLitros.setScale(2, RoundingMode.HALF_UP),
                volumenNetoTotal.setScale(2, RoundingMode.HALF_UP),
                porcentajeMermaGlobal,
                costoPromedioLitro,
                lotes.size()
        );

        // Consolidación por Producto
        Map<String, List<DesgloseLoteFinancieroDTO>> lotesPorProducto = desgloseLotes.stream()
                .collect(Collectors.groupingBy(DesgloseLoteFinancieroDTO::getProductoNombre));

        List<Producto> productosCalculo = productoId != null
                ? productos.stream().filter(p -> p.getId().equals(productoId)).collect(Collectors.toList())
                : productos;

        List<DesgloseProductoFinancieroDTO> desgloseProductos = productosCalculo.stream().map(prod -> {
            List<DesgloseLoteFinancieroDTO> lotesProd = lotesPorProducto.getOrDefault(prod.getNombre(), List.of());
            BigDecimal costoP = lotesProd.stream()
                    .map(DesgloseLoteFinancieroDTO::getCostoTotal)
                    .reduce(BigDecimal.ZERO, BigDecimal::add);
            BigDecimal volP = lotesProd.stream()
                    .map(DesgloseLoteFinancieroDTO::getVolumenNeto)
                    .reduce(BigDecimal.ZERO, BigDecimal::add);
            int unidadesP = lotesProd.stream()
                    .mapToInt(DesgloseLoteFinancieroDTO::getUnidadesEstimadas)
                    .sum();

            BigDecimal ingresoP = prod.getPrecio().multiply(BigDecimal.valueOf(unidadesP));
            BigDecimal gananciaP = ingresoP.subtract(costoP);
            BigDecimal margenP = ingresoP.compareTo(BigDecimal.ZERO) > 0
                    ? gananciaP.multiply(BigDecimal.valueOf(100)).divide(ingresoP, 1, RoundingMode.HALF_UP)
                    : BigDecimal.ZERO;

            return new DesgloseProductoFinancieroDTO(
                    prod.getId(),
                    prod.getNombre(),
                    prod.getTipoBebida(),
                    prod.getPrecio().setScale(2, RoundingMode.HALF_UP),
                    lotesProd.size(),
                    costoP.setScale(2, RoundingMode.HALF_UP),
                    volP.setScale(2, RoundingMode.HALF_UP),
                    unidadesP,
                    margenP
            );
        }).collect(Collectors.toList());

        return new DashboardFinancieroDTO(kpis, desgloseLotes, desgloseProductos);
    }
}

File: prototipo/backend/src/modules/financiero/financieroService.js
import { memoryStore } from '../../config/db.js';

export async function getDashboardFinanciero(productorId) {
  const productos = memoryStore.productos.filter(p => p.productor_id === productorId);
  const prodIds = productos.map(p => p.id);

  const lotes = memoryStore.lotes.filter(l => prodIds.includes(l.producto_id));

  // Si no hay lotes registrados
  if (lotes.length === 0) {
    return {
      totales: {
        costo_total_acumulado: 0,
        volumen_total_litros: 0,
        merma_total_litros: 0,
        volumen_neto_litros: 0,
        merma_porcentaje_promedio: 0,
        costo_promedio_litro: 0,
        lotes_contabilizados: 0
      },
      desglose_lotes: [],
      desglose_productos: []
    };
  }

  let costoTotalAcumulado = 0;
  let volumenTotalLitros = 0;
  let mermaTotalLitros = 0;

  const desgloseLotes = lotes.map(l => {
    const prod = productos.find(p => p.id === l.producto_id);
    const vol = parseFloat(l.volumen_litros) || 0;
    const merma = parseFloat(l.merma_litros) || 0;
    const costo = parseFloat(l.costo_total) || 0;
    const volNeto = Math.max(0, vol - merma);

    costoTotalAcumulado += costo;
    volumenTotalLitros += vol;
    mermaTotalLitros += merma;

    const mermaPct = vol > 0 ? (merma / vol) * 100 : 0;
    const costoPorLitro = volNeto > 0 ? costo / volNeto : 0;

    // Estimación por presentación (ej. 750ml = 0.75 L, 330ml = 0.33 L)
    let litrosPorUnidad = 0.75;
    if (prod && prod.presentacion.includes('330')) litrosPorUnidad = 0.33;
    if (prod && prod.presentacion.includes('500')) litrosPorUnidad = 0.5;

    const unidadesObtenidas = litrosPorUnidad > 0 ? Math.floor(volNeto / litrosPorUnidad) : 0;
    const costoPorUnidad = unidadesObtenidas > 0 ? costo / unidadesObtenidas : 0;
    const precioVenta = prod ? parseFloat(prod.precio) : 0;
    const ingresoBrutoEstimado = unidadesObtenidas * precioVenta;
    const gananciaBrutaEstimada = ingresoBrutoEstimado - costo;
    const margenBrutoPct = ingresoBrutoEstimado > 0 ? (gananciaBrutaEstimada / ingresoBrutoEstimado) * 100 : 0;

    return {
      lote_id: l.id,
      codigo_trazabilidad: l.codigo_trazabilidad,
      producto_nombre: prod ? prod.nombre : 'Desconocido',
      presentacion: prod ? prod.presentacion : '',
      fecha_produccion: l.fecha_produccion,
      volumen_litros: vol,
      merma_litros: merma,
      merma_porcentaje: Number(mermaPct.toFixed(2)),
      volumen_neto: Number(volNeto.toFixed(2)),
      costo_total: costo,
      costo_por_litro: Number(costoPorLitro.toFixed(2)),
      unidades_estimadas: unidadesObtenidas,
      costo_por_unidad: Number(costoPorUnidad.toFixed(2)),
      precio_venta_unitario: precioVenta,
      margen_bruto_porcentaje: Number(margenBrutoPct.toFixed(1))
    };
  });

  const volumenNetoTotal = Math.max(0, volumenTotalLitros - mermaTotalLitros);
  const mermaPorcentajePromedio = volumenTotalLitros > 0 ? (mermaTotalLitros / volumenTotalLitros) * 100 : 0;
  const costoPromedioLitro = volumenNetoTotal > 0 ? costoTotalAcumulado / volumenNetoTotal : 0;

  // Agrupación por producto
  const desgloseProductos = productos.map(p => {
    const lotesProd = desgloseLotes.filter(l => l.producto_nombre === p.nombre);
    const costoP = lotesProd.reduce((acc, it) => acc + it.costo_total, 0);
    const volP = lotesProd.reduce((acc, it) => acc + it.volumen_neto, 0);
    const unidadesP = lotesProd.reduce((acc, it) => acc + it.unidades_estimadas, 0);
    const ingresoP = unidadesP * p.precio;
    const gananciaP = ingresoP - costoP;
    const margenP = ingresoP > 0 ? (gananciaP / ingresoP) * 100 : 0;

    return {
      producto_id: p.id,
      nombre: p.nombre,
      tipo_bebida: p.tipo_bebida,
      precio: p.precio,
      lotes_count: lotesProd.length,
      costo_acumulado: costoP,
      volumen_neto_total: volP,
      unidades_estimadas_total: unidadesP,
      margen_promedio_porcentaje: Number(margenP.toFixed(1))
    };
  });

  const totalesObj = {
    costo_total_acumulado: Number(costoTotalAcumulado.toFixed(2)),
    volumen_total_litros: Number(volumenTotalLitros.toFixed(2)),
    merma_total_litros: Number(mermaTotalLitros.toFixed(2)),
    volumen_neto_litros: Number(volumenNetoTotal.toFixed(2)),
    porcentaje_merma_global: Number(mermaPorcentajePromedio.toFixed(2)),
    merma_porcentaje_promedio: Number(mermaPorcentajePromedio.toFixed(2)),
    costo_promedio_por_litro: Number(costoPromedioLitro.toFixed(2)),
    costo_promedio_litro: Number(costoPromedioLitro.toFixed(2)),
    lotes_contabilizados: lotes.length
  };

  return {
    kpis: totalesObj,
    totales: totalesObj,
    desglose_por_lote: desgloseLotes,
    desglose_lotes: desgloseLotes,
    desglose_productos: desgloseProductos
  };
}
